substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/UniqueShortNameProvider.java

```diff
@@ -0,0 +1,163 @@
+
+    public static synchronized UniqueShortNameProvider provider() {
+        if (provider == null) {
+            if (ImageSingletons.contains(UniqueShortNameProvider.class)) {
+                provider = ImageSingletons.lookup(UniqueShortNameProvider.class);
+            } else {
+                provider = new DefaultUniqueShortNameProvider();
+                ImageSingletons.add(UniqueShortNameProvider.class, provider);
+            }
+        }
+        return provider;
+    }
+
+    /**
+     * Returns a short, reasonably descriptive, but still unique name for the provided method
+     * which can be used as a linker local symbol.
+     */
+    public abstract String uniqueShortName(ResolvedJavaMethod method);
+    /**
+     * Returns a short, reasonably descriptive, but still unique name for a method as characterized
+     * by its loader name, owner class, method selector, signature and status as a method proper or
+     * a constructor.
+     */
+    public abstract String uniqueShortName(String loaderName, ResolvedJavaType declaringClass, String methodName, Signature methodSignature, boolean isConstructor);
+
+    /**
+     * Returns a short, reasonably descriptive, but still unique name for the provided
+     * {@link Method}, {@link Constructor}, or {@link Field}.
+     */
+    public abstract String uniqueShortName(Member m);
+
+    /**
+     * Returns a short, reasonably descriptive, but still unique name derived from the provided method
+     * that can be used as the name of a stub method defined on some factory class that invokes it.
+     * Note that although the value returned by this method is unique to the called method it's purpose
+     * is to uniquely define a single component of the full name of the stub method i.e. the method
+     * selector.
+     */
+
+    public abstract String uniqueStubName(ResolvedJavaMethod method);
+    /**
+     * Returns a unique identifier for a class loader that can be folded into the unique short name
+     * of methods where needed in order to disambiguate name collisions that can arise when the same
+     * class bytecode is loaded by more than one loader. For a few special loaders, such as the JDK
+     * runtime's builtin loaders and GraalVM Native's top level loaders an empty string is returned.
+     * There is no need to qualify the method name in this case since the method defined via these
+     * loaders is taken to be the primary instance. This is safe so long as the delegation model for
+     * this special set of loaders ensures that none of them will replicate a class loaded by another
+     * loader in the set.
+     * @param loader The loader whose identifier is to be returned.
+     * @return A unique identifier for the classloader or the empty string when the loader is one
+     * of the special set whose method names do not need qualification.
+     */
+    public abstract String classLoaderNameAndId(ClassLoader loader);
```

Comment:
If we really want this customizable then we should instead adjust the interface method parameter and use
```
String uniqueShortName(ClassLoader loader, ResolvedJavaType declaringClass, String methodName, Signature methodSignature, boolean isConstructor);
```
instead of
```
String uniqueShortName(String loaderName, ResolvedJavaType declaringClass, String methodName, Signature methodSignature, boolean isConstructor);
```
and then the `classLoaderNameAndId` implementation and its used fields are just private within `DefaultUniqueShortNameProvider`.

Reply:
I had to keep `classLoaderNameAndId` available as an API in `SubstrateUtil` because it is used by the method name comparison code. So, I moved the implementation back there. It does not need to appear as an overrideable API in `UniqueShortNameProvider`. A provider implementation can implement its own internal method for generating loader names, calling out to  `SubstrateUtil` when it wants the field value. The default provide can always just use the `SubstrateUtil` implementation.